### DISCUSSION TITLE
Handling inconsistencies for the Product Rating and Product Reviews blocks

### DISCUSSION THREAD

**Discussion by nefeline** [woocommerce/woocommerce#44967] [WooCommerce, Blocks, and Site Editing]
Currently, within WooCommerce blocks, we have the following blocks:

* Product Rating
* Product Details (which includes the Product Reviews section)
* Product Reviews

While separated, the product Rating and Reviews blocks have an implicit relationship between them that is not entirely supported. All current use cases are based on the assumption that product ratings are enabled for a given store and that users are using both the Product Rating and the Product Reviews blocks within the Single Product template/block, but that's not always the case, because:

* Users can remove the Product Reviews block and keep the Product Rating block within the Single Product Template, which creates an inconsistency as users can visualize average reviews but cannot read them individually or add new ones.
* Users can disable Product Reviews via `WooCommerce > Settings > Products` and still have the pre-existing ratings displayed within the Product Rating block regardless.

With that said, this is a discussion to gather feedback from folks regarding the following topics:

### 1. Should we respect the settings the user configured on `WooCommerce > Settings > Products`? What should happen if someone adds one of these blocks in the editor, but product reviews are disabled for their store?

**Proposed Solution:** 

* Allow the insertion of Product Rating and Product Reviews blocks in the editor independent of the selected settings on the backend. If the setting is disabled:
  *  **In the Editor:** a notice is displayed informing that in order to visualize the product reviews and ratings, the store owner has to enable the setting on `WooCommerce > Settings > Products`. 
  * **On the front end**: no content is rendered for the Product Rating and Product Reviews blocks.

### 2. What should happen if the store owner has the Product Reviews enabled on the backend but removes the Product Reviews block while preserving the Product Ratings within the Single Product template?

**Proposed Solution:**

* Ensure both the Product Reviews and Product Rating blocks share the same context so they can be aware of each other's existence within the editor.
* If the Product Reviews block is missing:
  * Add a notice to the Product Rating block informing the store owner that while users can visualize the average star rating, they won't be able to read the individual reviews or add new ones unless they add the corresponding Product Reviews block to the template.
  * Update the content of the Product Rating block, ensuring the "(X Customer reviews)" text doesn't link to the nonexistent reviews or, alternatively, altogether remove/disable the text.

* Allow users to hide/display the "(X Customer reviews)" text in the editor.
* Allow users to edit the link to the "(X Customer reviews)" text (in case they wanna link to a post/page elsewhere: e.g. they added the Product Reviews block on a separate/different post/page).

### 3. Are there recommended ways for sharing context between different blocks, independent of their position in the editor?

* In case we agree with the proposed solution in section 2: is there a specific guideline we should follow when it comes to ensuring that blocks share the same context and know about each other's existence in the editor? Are there any specific examples we can use? I can think of a few approaches, such as inner blocks and perhaps others that are a bit more "hacky", but I wanted to hear folks' opinions before making a decision/picking one.

Looking forward to hearing folks' thoughts on this =)

cc: @nerrad @albarin @roykho @gigitux @thealexandrelara @tjcafferkey @danielwrobert

**Comment by roykho** [comment]
Thanks for putting this together and sharing your thoughts on it.

> 1. Should we respect the settings the user configured on WooCommerce > Settings > Products? What should happen if someone adds one of these blocks in the editor, but product reviews are disabled for their store?

I think we probably need to preserve and retain the functionality of `WooCommerce > Settings > Products > Reviews` settings because it is not just a simple `enable/disable` as it has many other settings like `Enable star rating on reviews` and others making this section important. Though ideally, except for the select few settings, should not have the `enable/disable` part as that can be controlled via blocks. So my thought here is aligned with what your suggestion to show notices in the editor when working with these blocks that they need to enable them first in `WooCommerce > Settings > Products > Reviews`.

> 2. What should happen if the store owner has the Product Reviews enabled on the backend but removes the Product Reviews block while preserving the Product Ratings within the Single Product template?

Absolutely. They definitely need to be context aware of each other. Allowing users to show/hide the ratings and to be able to custom link the reviews to elsewhere is a good idea.

> 3. Are there recommended ways for sharing context between different blocks, independent of their position in the editor?

My knowledge here is limited and the only way I know is to just tap into `wp.data` to get the information we need from different blocks.

**Reply by nefeline** [reply]
Thank you so much for your input @roykho !

> I think we probably need to preserve and retain the functionality of WooCommerce > Settings > Products > Reviews settings because it is not just a simple enable/disable as it has many other settings like Enable star rating on reviews and others making this section important. Though ideally, except for the select few settings, should not have the enable/disable part as that can be controlled via blocks. So my thought here is aligned with what your suggestion to show notices in the editor when working with these blocks that they need to enable them first in WooCommerce > Settings > Products > Reviews.

💯 

> Absolutely. They definitely need to be context aware of each other. Allowing users to show/hide the ratings and to be able to custom link the reviews to elsewhere is a good idea.

👍 

> My knowledge here is limited and the only way I know is to just tap into wp.data to get the information we need from different blocks.

Thanks! Yeah, same here: I do have a few ideas to get this done, but if there's a Gutenberg-recommended approach to handle this scenario, I would love to hear more about it.

**Comment by nerrad** [comment]
> Should we respect the settings the user configured on WooCommerce > Settings > Products? What should happen if someone adds one of these blocks in the editor, but product reviews are disabled for their store?

This is one of those things that I think lends itself to a UI/UX pattern that needs to be applied consistently across Woo as we transition to the block paradigm. ccing @jarekmorawski, @elizaan36, @chrisdesrochers, @pmcpinto, and @joshuatf  for awareness of this discussion since I think this particular item is part of a larger conversation around settings I/A in context with relevant blocks. 

### Long term

Below are some suggested principles we should follow for all blocks including how it could apply to this specific discussion. I think we should create an issue to address this in subsequent iterations.

#### Global settings are for _defaults_
So for example, in this context, the options "Enable star rating on reviews", "Star ratings should be required, not optional", "Show 'verified owner' label on customer reviews" and "Reviews can only be left by 'verified owners'" would be the defaults for _all_ review block instances but can still be modified at the block level. 

Whereas "Enable product reviews" is an option that would be shifted to be controlled by the block's presence in the template and not appear as a global setting.

#### Where there are global defaults, block either link to those or offer the ability to change the default in place
Users shouldn't have to search through a bunch of settings to find out how to change the defaults.

### Short term

To solve the immediate problem:

- Implement your suggestion for the block behaviour dependent on "Enable product reviews".
- Use the other review and rating global settings to control the default for a block when it is inserted and doesn't have relevant attributes set.

> What should happen if the store owner has the Product Reviews enabled on the backend but removes the Product Reviews block while preserving the Product Ratings within the Single Product template?

- I don't think they should be aware of the presence of each other, for now. In the future, I think we should offer an option for leaving reviews immediately (via a modal).
- The link from the ratings to the reviews should definitely be something that can be toggled as _well_ as customizable. 

> Are there recommended ways for sharing context between different blocks, independent of their position in the editor?
Options are:

- `wp.data` (via [`useEntityRecord`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-core-data/#useentityrecord)) for sharing across any part of the view. 
- [Block Context](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-context/) for within a block tree. 

This is something I've been thinking about as a part of the work on the price block refinements and I think we should introduce an interface that roughly looks like this:

```js
const useProductEntity = ( { context } ) => {
    if ( context?.productId || context ?.postId ) {
         const productId = context?.productId || context?.postId;
         // ... implement useEntityRecord to get the product and return it 
    }
    // there is no context provided so just implement `useEntityRecord` and return the product.
}
```
** Note, the implementation of `useEntityRecord` would be depending on the outcome of [this discussion](https://github.com/woocommerce/woocommerce-blocks/discussions/9993).

Then every block that requires products would use that hook and if there is block context provided through the tree - that gets used. If the block is outside of a tree without context, then it's pulled from the `wp.data` store for the product entity.

I've seen a mixture of methods used across our blocks and I think we should move to a consistent implementation.

**Reply by nerrad** [reply]
Actually, scratch my proposed interface above for `useProductEntity`. It wouldn't work. I forgot the rule of hooks about conditions and there also is still the need for the hook to know what id to use if context isn't provided.

The main point though is that I think whatever we implement for grabbing product related data should be consistent (and I think [the discussion here](https://github.com/woocommerce/woocommerce-blocks/discussions/9993) will help inform that.

**Reply by nefeline** [reply]
Thanks for the feedback & guidance @nerrad, much appreciated!

Before diving into the short and long-term solutions, I'd like to share my thoughts on this as well: from a macro perspective, I believe we have 3 different "ranks" of settings:

1. **Functional settings:** These are the ones where the user can decide to turn on or off a given functionality. In our use case here, "Enable product reviews" could be categorized as a core setting.
2. **Style and layout settings:** Any setting that controls how a given feature should be displayed on the FE can be part of this rank. In our use-case here we have: "Enable star rating on reviews" and "Show "verified owner" label on customer reviews". 
3. **Permission settings:** Any setting that controls who should have access to a given feature. The "Reviews can only be left by "verified owners"" setting fit this criteria.

With that said, I believe that, in the


## Long term

- If a given block has a **Functional setting** (rank 1), it can be deprecated. E.g., considering we have a block for product reviews, there's no need to maintain a setting to enable/disable it. I believe this is aligned with what you mentioned here:

> Whereas "Enable product reviews" is an option that would be shifted to be controlled by the block's presence in the template and not appear as a global setting.

- If a given block has a setting that fits the **Style and layout** (rank 2), create the equivalent of it on the Block Editor and deprecate the global setting. IMO these settings could be per-block **exclusively**. If I'm using the rating block in 2 different areas of my website, and I want to see stars in one place, while having plain text in another, I should be able to do so without any problems. In this particular case, I believe it's safe to abandon/deprecate any pre-existing global settings completely.

With that said, regarding the following:

> Global settings are for defaults

I wouldn't maintain any connection between global and per-block settings in this particular case (for settings "rank 2"), as I believe styling should be controlled by the editor exclusively; in other words, I agree with @elizaan36 that:

> This creates a situation where the block settings have different configuration from admin settings, which is super confusing.

So giving styling/layout settings a new home, per block, on the editor side, without any connection with the global settings, is my suggestion.

- If a given block has a setting that is classified as a "permission setting" (rank 3), preserve it as a global setting: if I want to allow only verified owners to leave product reviews on my website, this is the type of setting that IMO doesn't make much sense to be changed per block.

So regarding:

> Where there are global defaults, block either link to those or offer the ability to change the default in place

I agree with the idea of linking to those or offering the ability to change the default in place while also being very explicit that this setting will affect all relevant blocks universally and not just the one where the setting is being changed.

Now, regarding the:

## Short term

👍 I agree with the proposed path, makes sense. cc: @albarin FYI and in case you wanna chime in since we are both dividing and conquering the short-term optimization for this block :)

Regarding the:

## Recommended ways for sharing context between different blocks

> the implementation of useEntityRecord would be depending on the outcome of https://github.com/woocommerce/woocommerce-blocks/discussions/9993.

👍

> I think we should create an issue to address this in subsequent iterations.

👍  I'm creating an issue to capture the outcome of this discussion in the long term.

**Reply by nerrad** [reply]
> I wouldn't maintain any connection between global and per-block settings in this particular case (for settings "rank 2"), as I believe styling should be controlled by the editor exclusively; in other words, I agree with @elizaan36 that:

I think that's fine as long as there is some other mechanism for choosing what blocks will start with every time they are inserted (which WP global styles support can come in).

**Reply by albarin** [reply]
I agree with the short-term proposed path. I think what we are implementing goes in line with it as well (splitting the Product Reviews into stars & link blocks).

> I don't think they should be aware of the presence of each other, for now. In the future, I think we should offer an option for leaving reviews immediately (via a modal).

> The link from the ratings to the reviews should definitely be something that can be toggled as well as customizable.

Agree, feels weird to me to have two blocks depend on each other like that. We can allow to customize/toggle the link in this new _review link_ block.

**Reply by joshuatf** [reply]
@nerrad I have also previously proposed a `useProductEntity` hook and still like this idea as a layer of abstraction between blocks and the need to understand their context explicitly.

I think we can still use your proposed hook with some minor tweaks:

```
const useProductEntity = ( { context } ) => {
    const productId = context?.productId || context?.postId || useEntityId( 'postType', 'product' );
    const product = useEntityRecord( 'postType', 'product', productId );
    return product;
}
```

I'll need to explore what precedence each context/entity provider should take, but that should satisfy the basic needs of each context being able to obtain its product and establish consistency of the shape of the data by using `useEntityRecord`.  I think this issue also really reenforces why we shouldn't be overriding the entity REST endpoint and changing the shape of data as we discussed in https://github.com/woocommerce/woocommerce-blocks/discussions/9993.

One thing the above hook does not address is whether or not we're getting the provided entity or the edited state of the entity, which is currently more specific to editor.  The above hook would not work for the reasons you mentioned about conditions above, but we can create a separate hook `useEditedProductEntity` to be more explicit about which record we're grabbing as I know this has already created some confusion for @woocommerce/mothra as to which `product` entity we're currently working with.

**Comment by elizaan36** [comment]
> So for example, in this context, the options "Enable star rating on reviews", "Star ratings should be required, not optional", "Show 'verified owner' label on customer reviews" and "Reviews can only be left by 'verified owners'" would be the defaults for all review block instances but can still be modified at the block level.

This creates a situation where the block settings have different configuration from admin settings, which is super confusing. I agree there should be a standard approach but think the block settings and admin settings should be consistent. The merchant may change the setting on blocks and think it will be updated for the store.

**Reply by nerrad** [reply]
This seems like something good to check via user testing. There are a number of assumptions I've seen when it comes to block settings and "global" settings and to a degree the behaviour is also influenced by user familiarity with blocks and what learned patterns they may already have from that usage.

**Reply by joshuatf** [reply]
Walking the line between seasoned users (those more familiar with traditional WooCommerce settings) and newly onboarded users can be tricky.

I agree that divergence between the global setting and the presence of the block in the template is confusing and doesn't make sense.  Another option we have aside from showing a message that product reviews are disabled is to auto-insert or remove the block from the template when the global setting is toggled and similarly, toggling the global setting when the block is added or removed from templates.  This should be trivial in the near future with the auto-insertion feature being experimented within Gutenberg.

I'm not sure that this is less confusing from a UX perspective though it does remove the need for:
* Wordy notices in a block render explaining why the recently added block won't work.
* Further action on the merchant's part to allow this block to work.
* Removing the global setting eventually leaving more seasoned users questioning how to toggle reviews on/off.

**Reply by joshuatf** [reply]
On reading again, I think @nerrad is proposing a similar idea in his long-term solution:

>Where there are global defaults, block either link to those or offer the ability to change the default in place

**Reply by elizaan36** [reply]
Agreed these flows need testing for sure to understand user expectations. In the meantime, this proposal makes a lot of sense:

> auto-insert or remove the block from the template when the global setting is toggled and similarly, toggling the global setting when the block is added or removed from templates.

If this is possible, it would prevent the editor being out of sync with admin settings and would align with the intention of the user to add or remove the block. When looking into the details of the flows there may be some refining needed but overall seems to simplify things.

**Reply by jarekmorawski** [reply]
I second this. There should be a connection between global settings and block. After all, blocks represent the store's internal configuration, and any disambiguation between them may lead to confusion.

We should, however, figure out a way to communicate that to users. I'm thinking of a small note in the inserter mentioning the link between enabled features and possible missing blocks. Otherwise, we risk that the user may believe that some features do not exist in Woo after they were handed over a store from someone else who set up the backend.
